### PR TITLE
Create import YAML page

### PR DESCRIPTION
**Pull request by nicolethoen**
Here is the page as it stands ATM.
I put the link to Create from YAML under the Home navigation, because I'm not sure it makes sense to have a project selector at the top of a generic uploader. The namespace is specified in the YAML and if they have one project selected, but then paste some YAML for a different namespace, I'm not sure what the behavior should be. 

screen shot 2018-09-18 at 10 58 58 am

### CODE REVIEW

File: frontend/public/components/importYaml.jsx [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+  constructor(props) {
+    super(props);
+  }
+
+  render() {
+    const match = {url: '', params: {ns: 'default', plural: 'pods'}, isExact: true, path: ''};
+    return <React.Fragment>
+      <div className="yaml-editor-header">
+        <div>Import YAML</div>
+        <div className="yaml-editor-subheader">Create or replace resources from their YAML or JSON definitions. Use the template below or browse to upload a file.</div>
```
Comment (jhadvig) [inline frontend/public/components/importYaml.jsx]: Do we want this component for resource replacement ? If so then `Create from YAML` is not accurate. maybe `Import YAML` ?

File: frontend/public/components/create-yaml.tsx [hunk 2]
```diff
@@ -13,6 +14,7 @@ import { ClusterServiceVersionModel } from '../models';
 export const CreateYAML = connectToPlural((props: CreateYAMLProps) => {
   const {match, kindsInFlight, kindObj} = props;
   const {params} = match;
+  const showHeader = _.has(props, 'showHeader') ? props.showHeader : true;
```
Comment (jhadvig) [inline frontend/public/components/create-yaml.tsx]: maybe just
```js
_.get(props, 'showHeader', true)
```

File: frontend/public/components/importYaml.jsx (hunk 1 above)
Comment (nicolethoen) [reply frontend/public/components/importYaml.jsx]: It's possible. I pulled that wording from the current suggested design, but all of that is not ironed out yet.

File: frontend/public/components/import-yaml.jsx [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+  constructor(props) {
+    super(props);
+  }
+
+  render() {
+    const match = {url: '', params: {ns: 'default', plural: 'pods'}, isExact: true, path: ''};
+    return <React.Fragment>
+      <div className="yaml-editor-header">
+        <div>Import YAML</div>
+        <div className="yaml-editor-subheader">Create or replace resources from their YAML or JSON definitions. Use the template below or browse to upload a file.</div>
```
Comment (jhadvig) [inline frontend/public/components/import-yaml.jsx]: Also not really sure what the `... browse to upload a file` should mean, since there is no `Browse` button.

File: frontend/public/components/_edit-yaml.scss [hunk 4]
```diff
@@ -1,9 +1,11 @@
 .yaml-editor-header {
   border-bottom: 1px solid #ddd;
   font-size: 16px;
-  height: 45px;
   padding: 10px 20px;
 }
+.yaml-editor-subheader {
```
Comment (spadgett) [inline frontend/public/components/_edit-yaml.scss]: I realize the existing names aren't really following BEM, but this should probably be `.yaml-editor__subheader` (and existing class above `.yaml-editor__header`)

http://getbem.com/naming/

cc @rhamilto

File: frontend/public/components/importYaml.jsx (hunk 1 above)
Comment (jhadvig) [reply frontend/public/components/importYaml.jsx]: Ok, also not really sure what the `... browse to upload a file` should mean, since there is no `Browse` button.

**Comment by jhadvig** [comment]
@nicolethoen we should make the `Donwload` button optional here, since all the input will be supplied by the user anyway.

File: frontend/public/components/import-yaml.jsx [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+      </div>
+      <CreateYAML match={match} kindsInFlight={false} kindObj={PodModel} showHeader={false} />
```
Comment (spadgett) [inline frontend/public/components/import-yaml.jsx]: `PodModel` doesn't seem right... It would be better to make that an optional prop and not specify it for this case.

**Comment by robszumski** [comment]
Linking https://github.com/openshift/openshift-origin-design/pull/108 since it has discussion of this topic.

> The namespace is specified in the YAML and if they have one project selected, but then paste some YAML for a different namespace, I'm not sure what the behavior should be.

I guess the options are throw up a warning, or just create it. I believe that an error is thrown if you do `kubectl create -n foo -f bar.yaml` and the yaml has a different namespace in it. Seems like reasonable behavior to me...buuut if you did want to use the namespace in the file, is there a way we could include a way to use it?

"The current namespace does not match the uploaded file. You can [use the current namespace]() or [use foobar namespace]()."

File: frontend/public/components/import-yaml.jsx (hunk 5 above)
Comment (nicolethoen) [reply frontend/public/components/import-yaml.jsx]: If it is not specified, then there would be nothing in the editor, when it loads.
It doesn't seem like that was a state in any of the designs I was seeing. Does an empty editor give enough cues to the user?

Comment (spadgett) [reply frontend/public/components/import-yaml.jsx]: Presumably here you already have the YAML you want to use and don't need to work from a template. Having content prefilled might even get in the way since you have to select or delete it to paste something new.

We had some designs for drag n drop support where there's an empty state message saying to drag or paste a file. We don't necessarily need that for the first pass, though.

**Comment by nicolethoen** [comment]
Here is a screenshot of the latest appearance after the latest round of changes.

screen shot 2018-09-21 at 3 37 26 pm

File: frontend/public/components/create-yaml.tsx [hunk 6]
```diff
@@ -7,18 +7,17 @@ import { connectToPlural } from '../kinds';
...
-  const {match, kindsInFlight, kindObj} = props;
+  const {match, kindsInFlight, kindObj, download = true, showHeader = true} = props;
   const {params} = match;
 
   if (!kindObj) {
     if (kindsInFlight) {
       return <LoadingBox />;
     }
-    return <ErrorPage404 />;
+    return <AsyncComponent loader={() => import('./edit-yaml').then(c => c.EditYAML)} create={true} showHeader={showHeader} download={download} />;
```
Comment (spadgett) [inline frontend/public/components/create-yaml.tsx]: I think it's better to just create a separate ImportYAML component since almost all the logic is in the EditYAML component anyway.

File: frontend/public/components/edit-yaml.jsx [hunk 7]
```diff
@@ -159,12 +159,12 @@ export const EditYAML = connect(stateToProps)(
...
         this.doc = es.getDocument();
       }
-      let yaml;
+      let yaml = '';
 
       try {
         yaml = safeDump(obj);
       } catch (e) {
-        yaml = `Error dumping YAML: ${e}`;
+        //yaml = `Error dumping YAML: ${e}`;
```
Comment (spadgett) [inline frontend/public/components/edit-yaml.jsx]: We should only try to `safeDump` if there is an object.

File: frontend/public/components/edit-yaml.jsx [hunk 8]
```diff
@@ -283,31 +283,28 @@ export const EditYAML = connect(stateToProps)(
...
         Rendering:
         Our parent divs are meta objects created by third parties... but we need 100% height in all parents for flexbox :-/
         The current solution uses divs that are relative -> absolute -> flexbox pinning the button row with margin-top: auto
       */
 
       const {error, success, stale} = this.state;
-      const {create, obj, showHeader = false, readOnly} = this.props;
-      const kind = obj.kind;
+      const {create, obj, download = true, showHeader = false, readOnly} = this.props;
+      const kind = obj ? obj.kind : null;
```
Comment (spadgett) [inline frontend/public/components/edit-yaml.jsx]:
```js
const kind = obj && obj.kind;
```

File: frontend/public/components/import-yaml.tsx [hunk 9]
```diff
@@ -0,0 +1,17 @@
+import * as React from 'react';
+import { CreateYAML } from './create-yaml';
+
+export class ImportYamlPage extends React.Component {
```
Comment (spadgett) [inline frontend/public/components/import-yaml.tsx]: Make this a stateless functional component.

https://medium.com/@ethan_ikt/react-stateless-functional-component-with-typescript-ce5043466011

File: frontend/public/components/namespace.jsx [hunk 10]
```diff
@@ -275,6 +275,10 @@ class NamespaceDropdown_ extends React.Component {
         defaultBookmarks={defaultBookmarks}
         storageKey={NAMESPACE_LOCAL_STORAGE_KEY}
         shortCut="n" />
+      <div className="co-create-from-yaml co-m-nav-link">
+        <span className="pficon pficon-add-circle-o"></span>
```
Comment (spadgett) [inline frontend/public/components/namespace.jsx]: aria-hidden="true"

File: frontend/public/components/import-yaml.tsx [hunk 11]
```diff
@@ -0,0 +1,17 @@
...
+
+  render() {
+    const match = {url: '', params: {ns: 'default', plural: ''}, isExact: true, path: ''};
+    return <React.Fragment>
+      <div className="yaml-editor__header">
+        <div>Import YAML</div>
+        <div className="yaml-editor__subheader">Create or replace resources from their YAML or JSON definitions. Use the template below or browse to upload a file.</div>
```
Comment (spadgett) [inline frontend/public/components/import-yaml.tsx]: We don't support replace or file upload yet.

File: frontend/public/components/create-yaml.tsx (hunk 6 above)
Comment (spadgett) [inline frontend/public/components/create-yaml.tsx]: You should be able to undo this change now, correct?

File: frontend/public/components/edit-yaml.jsx [hunk 12]
```diff
@@ -72,6 +72,9 @@ export const EditYAML = connect(stateToProps)(
     }
 
     getModel(obj) {
+      if (_.isEmpty(obj)) {
+        return;
```
Comment (spadgett) [inline frontend/public/components/edit-yaml.jsx]: return null;

File: frontend/public/components/namespace.jsx [hunk 13]
```diff
@@ -275,6 +275,10 @@ class NamespaceDropdown_ extends React.Component {
         defaultBookmarks={defaultBookmarks}
         storageKey={NAMESPACE_LOCAL_STORAGE_KEY}
         shortCut="n" />
+      <div className="co-create-from-yaml co-m-nav-link">
+        <span className="pficon pficon-add-circle-o co-create-from-yaml__add-icon" aria-hidden="true"></span>
+        <a href={formatNamespacedRouteForResource('import', activeNamespace)}>Create from YAML</a>
```
Comment (spadgett) [inline frontend/public/components/namespace.jsx]: We have a different action name here than we show at the top of the page. Should it be "Create from YAML" or "Import YAML"?

File: frontend/public/components/_dropdown.scss [hunk 14]
```diff
@@ -162,6 +162,18 @@ $color-bookmarker: #DDD;
...
   }
+
+  .co-create-from-yaml {
+    margin-left: auto;
+
+    @media (min-width: $grid-float-breakpoint) {
+      padding-top: 7px;
+    }
+
+    .co-create-from-yaml__add-icon {
```
Comment (spadgett) [inline frontend/public/components/_dropdown.scss]: nit: I would avoid nesting this rule since it's not necessary with BEM naming

File: frontend/public/components/import-yaml.tsx [hunk 15]
```diff
@@ -0,0 +1,25 @@
...
+    </div>
+    <AsyncComponent loader={() => import('./edit-yaml').then(c => c.EditYAML)} showHeader={props.showHeader} download={props.download} />;
+  </React.Fragment>;
+
+};
+
+/* eslint-disable no-undef, no-unused-vars */
+export type ImportYamlProps = {
+  showHeader?: boolean;
+  download?: boolean;
```
Comment (spadgett) [inline frontend/public/components/import-yaml.tsx]: I don't think we need these props since they should always be false?

File: frontend/public/components/create-yaml.tsx [hunk 16]
```diff
@@ -7,18 +7,18 @@ import { connectToPlural } from '../kinds';
...
-  const {match, kindsInFlight, kindObj} = props;
+  const {match, kindsInFlight, kindObj, showHeader = true} = props;
   const {params} = match;
 
   if (!kindObj) {
     if (kindsInFlight) {
       return <LoadingBox />;
     }
-    return <ErrorPage404 />;
+    <ErrorPage404 />;
```
Comment (spadgett) [inline frontend/public/components/create-yaml.tsx]: missing return

File: frontend/public/components/namespace.jsx [hunk 17]
```diff
@@ -275,6 +275,11 @@ class NamespaceDropdown_ extends React.Component {
         defaultBookmarks={defaultBookmarks}
         storageKey={NAMESPACE_LOCAL_STORAGE_KEY}
         shortCut="n" />
+      <div className="co-create-from-yaml co-m-nav-link">
```
Comment (spadgett) [inline frontend/public/components/namespace.jsx]: This isn't really a nav-link, so the class name is not really accurate.

I'd suggest `co-import-yaml-link` as well.

File: frontend/public/components/_dropdown.scss [hunk 18]
```diff
@@ -164,6 +164,22 @@ $color-bookmarker: #DDD;
       white-space: nowrap;
     }
   }
+
+  .co-create-from-yaml {
+    margin-left: auto;
+
+    &:hover {
+      background-color: transparent !important;
```
Comment (spadgett) [inline frontend/public/components/_dropdown.scss]: Is this needed if we don't use the `co-m-nav-link` style? It might be better to have our own `co-m-masthead-link` or similar

File: frontend/public/components/edit-yaml.jsx [hunk 19]
```diff
@@ -290,31 +294,28 @@ export const EditYAML = connect(stateToProps)(
     }
 
     render () {
-      if (_.isEmpty(this.props.obj)) {
```
Comment (spadgett) [inline frontend/public/components/edit-yaml.jsx]: Maybe leave this with a check for edit?

```js
if (!this.props.create && !this.props.obj) {
  return <Loading />;
}
```

File: frontend/public/components/edit-yaml.jsx [hunk 20]
```diff
@@ -160,13 +160,16 @@ export const EditYAML = connect(stateToProps)(
...
         this.doc = es.getDocument();
       }
-      let yaml;
+      let yaml = '';
 
-      try {
-        yaml = safeDump(obj);
-      } catch (e) {
-        yaml = `Error dumping YAML: ${e}`;
+      if (!_.isEmpty(obj)) {
```
Comment (spadgett) [inline frontend/public/components/edit-yaml.jsx]: Is `obj` ever set to `{}`? Otherwise, it can be simple

```js
if (obj) {
  // ...
}
```

File: frontend/public/components/_dropdown.scss [hunk 21]
```diff
@@ -166,6 +169,51 @@ $color-bookmarker: #DDD;
...
+  &.active {
+    padding-left: $link-padding-left - $active-border;
+    background-color: #383F47;
+    border-left: $active-border solid $color-os-nav-border-left-active;
+
+    a {
+      color: #fff;
+    }
+  }
+}
```
Comment (spadgett) [inline frontend/public/components/_dropdown.scss]: @rhamilto Would you mind reviewing the styles here?

I suggested a `co-m-masthead-link` class since `co-m-nav-link` seemed wrong., but we might want to share the definition with `co-m-nav-link`

Comment (rhamilto) [reply frontend/public/components/_dropdown.scss]: This looks like a straight duplication of [`.co-m-nav-link`](https://github.com/openshift/console/blob/master/frontend/public/components/_nav.scss#L6-L30)?  I'd be inclined update _nav.scss with the shared stuff.

```
.co-m-masthead-link,
.co-m-nav-link {
  display: block;
  float: none;
  font-size: 13px;
  line-height: 23px;
  overflow-x: hidden;
  padding: 2px 0 2px $link-padding-left;
  white-space: nowrap;

  a {
    background-color: inherit;
    color: $color-os-nav-title;
    display: inline-block;
    text-decoration: none;
    width: 100%;
  }

  &.active {
    padding-left: $link-padding-left - $active-border;
    background-color: #383F47;
    border-left: $active-border solid $color-os-nav-border-left-active;
    a {
      color: #fff;
    }
  }
}

.co-m-nav-link {
  &:hover {
   background-color: $color-os-nav-background-active;
  }

  &:focus {
   background-color: transparent;
  }
}
```

File: frontend/public/components/_nav.scss [hunk 22]
```diff
@@ -3,7 +3,8 @@ $link-padding-left: 40px;
 $section-padding-left: 13px;
 $active-border: 3px;
 
-.co-m-nav-link {
+.co-m-nav-link,
+.co-m-masthead-link {
```
Comment (spadgett) [inline frontend/public/components/_nav.scss]: If you pull out these styles as @rhamilto recommends, you should be able to get rid of the `!important` override above.

```css
.co-m-nav-link {
  &:hover {
   background-color: $color-os-nav-background-active;
  }

  &:focus {
   background-color: transparent;
  }
}
```

See https://github.com/openshift/console/pull/540#discussion_r221079268

Comment (nicolethoen) [reply frontend/public/components/_nav.scss]: I am trying to make the hover behavior of the Import YAML link match that of the namespace selector. This behavior is different than the hover behavior of the side bar nav items. That is why the !important override is there.

Comment (spadgett) [reply frontend/public/components/_nav.scss]: But if you break up the rules as @rhamilto suggests, you shouldn't need `!important` because you won't have to override a value.

**Comment by nicolethoen** [comment]
jenkins rebuild

**Comment by spadgett** [comment]
CI will fail until we rebuild the shared cluster.

File: frontend/public/components/_dropdown.scss [hunk 23]
```diff
@@ -166,6 +166,22 @@ $color-bookmarker: #DDD;
   }
 }
 
+.co-import-yaml-link {
```
Comment (rhamilto) [inline frontend/public/components/_dropdown.scss]: This CSS addition should probably live somewhere other than _dropdown.scss since these elements are not dropdown related?

Comment (nicolethoen) [reply frontend/public/components/_dropdown.scss]: the link lives in the co-namespace-selector dropdown. Would it be better to create a new scss file for the link?

Comment (rhamilto) [reply frontend/public/components/_dropdown.scss]: That'd be my vote.  IMO, the styles in _dropdown.scss be specific to dropdowns.

**Comment by nicolethoen** [comment]
jenkins rebuild

**Comment by nicolethoen** [comment]
jenkins rebuild

File: frontend/integration-tests/tests/olm/prometheus.scenario.ts [hunk 24]
```diff
@@ -117,9 +117,9 @@ describe('Interacting with the Prometheus OCS', () => {
 
   it('displays the raw YAML for the `Prometheus`', async() => {
     await element(by.linkText('YAML')).click();
-    await browser.wait(until.presenceOf($('.yaml-editor--buttons')));
-    await $('.yaml-editor--buttons').element(by.buttonText('Save')).click();
-    await browser.wait(until.visibilityOf(crudView.successMessage), 1000);
+    await browser.wait(until.presenceOf($('.yaml-editor__buttons')));
+    await $('.yaml-editor__buttons').element(by.buttonText('Save')).click();
+    await browser.wait(until.visibilityOf($('.alert-success')), 1000);
```
Comment (spadgett) [inline frontend/integration-tests/tests/olm/prometheus.scenario.ts]: We should rollback the change to this line

Comment (nicolethoen) [reply frontend/integration-tests/tests/olm/prometheus.scenario.ts]: I can make the change when I am adding drag and drop
